### ISSUE TITLE
Conditional compilation using Requires,jl

### ISSUE THREAD

**Issue by AayushSabharwal** [JuliaDynamics/Agents.jl#397]
@Libbum pointed out [Requires.jl](https://github.com/JuliaPackaging/Requires.jl) which only loads parts of the package as and when required, based on dependency packages being loaded. Opening this issue to discuss what all can/should be conditionally loaded. Currently what I've thought of: 

- OSMSpace, depending on OpenStreetMapX.jl
- GraphSpace, depending on LightGraphs.jl
- Data collection functionality, depending on DataFrames.jl

All of this should probably be able to be put in `src/Agents.jl`, with some `@require` macros in an `__init__` function, as described in the Requires.jl repo README.

Am I missing something? Is there something I've interpreted incorrectly? Any other suggestions?

**Comment by Libbum** [comment]
This would be 'nice', but I don't think it's truly how Requires works. If we can though, let's think about how exactly this may be possible.

**How I understand it:**
Example: main file of [InteractiveDynamics](https://github.com/JuliaDynamics/InteractiveDynamics.jl/blob/master/src/InteractiveDynamics.jl). There are some base things loaded regardless of which other **packages** are loaded. But, since ID is becoming essentially a meta-package for visualisation in JuliaDynamics, there is extra functionality hidden behind the Requires guards. There's very little reason to load in Billiards files when all we want to do is look at Agents plots.

For Agents, we could currently put the `src/visualization` files behind a Requires guard triggering on `Plots` for example, but that's a very light aspect of the package. Not really worth the trouble.

**What is proposed above:**
In contrast, we'd like to only load OSMX.jl when users want to use `OSMSpace`. Since `OSMSpace` is not a package, my read here is that we can't actually use the Requires guard to trigger the package load. I could be wrong though...

**What may be possible:**
```julia
julia> using Agents
julia> space = OpenStreetMapSpace(TEST_MAP)
ERROR: UndefVarError: OpenStreetMapSpace not defined
julia> using Agents.MapSpace #i.e. some submodule
julia> space = OpenStreetMapSpace(TEST_MAP)
OpenStreetMapSpace with 1456 roadways and 1799 intersections
```
In other words, having submodules for certain features that require additional packages.

This however, is somewhat annoying. Its nice that everything is in one place at the moment. So we will need to consider how effective such a split will be&mdash;will this dramatically improve the time-to-compile and time-to-plot metrics at this stage?

With that being said; Agents has managed to stay pretty lean on dependencies for now, but as we add in some of the new planned features, some separation may be necessary. In that sense, it's timely to look into this, although it will probably be a breaking change and thus be a priority for v5, not the immediate future.

**Comment by AayushSabharwal** [comment]
Since most of the data collection relies on DataFrames.jl, is it worth putting that behind a `@require` too? I'm working on timing it

**Comment by Libbum** [comment]
It's not a big hog, but sure&mdash;if you can find a way to trigger it.

**Comment by AayushSabharwal** [comment]
You're right, there's basically no difference, so not worth

**Comment by Datseris** [comment]
Ummm, I don't think putting things behind require blocks should be done so lightly. Definitely I cannot see a reason that dataframes-related stuff should be behind requires blocks. Let's rewind a bit, and provide arguments why everything should be behind requires block. Because if one takes this too far, everything that uses something beyond base Julia could be behinds a requires block, and then to actually use Agents.jl in practice you need to start a `using` statement followed by 10 package names.

Plotting is an exceptionally special case that goes behind requires for the sole reason of download size and build speed, and almost exclusively because of Plots.jl. AbstractPlotting.jl does not have these problems and I wouldn't put it behind a requires statement if Agents.jl has already moved to Makie.jl. DataFrames downloads and builds instantly, there is no gain in requiring it specially. InteractiveDynamics.jl is also a special case that is not a package that you would use in a script in your "day-to-day" work. I would be very hesitant basing design decisions of normal packages on InteractiveDynamics.jl. The purpose of InteractiveDynamics.jl having things structured in requires blocks by the way is solely so that big libraries like DynamicalSystems.jl are not installed when you install the package, which are clearly frontend libraries.

I see DataFrames.jl, as well as LightGraphs.jl an integral part of Agents.jl, and I'd like to see arguments on why they should be behind requires blocks. Both are quite light packages. If the argument is reduction of compilation time, then I'd like to see timings of how large this reduction actually is.

**Comment by Libbum** [comment]
Agreed on all points. The only current dependency that's not integral but included is `OpenStreetMapX`, which contributes about 20% of the present load time. Would be nice if it wasn't so heavy, but without fundamentally changing (in a breaking manner)how we structure things, there's little we can do at this stage.

**Comment by Libbum** [comment]
Relevant discussion concerning this can be carried over in #458.
